src/vmm/src/devices/virtio/net/metrics.rs

```diff
@@ -0,0 +1,349 @@
+//! (i.e the number of times an API request failed). These metrics are reset upon flush.
+//! We use NET_DEV_METRICS_PVT instead of adding an entry of NetDeviceMetrics
+//! in Net so that metrics are accessible to be flushed even from signal handlers.
+
+use std::collections::BTreeMap;
+use std::sync::{RwLock, RwLockReadGuard};
+
+use serde::ser::SerializeMap;
+use serde::{Serialize, Serializer};
+
+use crate::logger::{IncMetric, SharedIncMetric};
+
+/// returns lock-unwrapped NET_DEV_METRICS_PVT.metrics,
+/// we have this function so that we don't have to make
+/// NET_DEV_METRICS_PVT public
+pub fn get_net_metrics() -> RwLockReadGuard<'static, NetMetricsPerDevice> {
+    // lock is always initialized so it is safe the unwrap the lock without a check.
+    NET_DEV_METRICS_PVT.read().unwrap()
+}
+
+/// map of network interface id and metrics
+/// this should be protected by a lock before accessing.
+#[derive(Debug)]
+pub struct NetMetricsPerDevice {
+    /// used to access per net device metrics
+    pub metrics: BTreeMap<String, NetDeviceMetrics>,
+}
+
+impl NetMetricsPerDevice {
+    /// Allocate `NetDeviceMetrics` for net device having
+    /// id `iface_id`. Also, allocate only if it doesn't
+    /// exist to avoid overwriting previously allocated data.
+    /// lock is always initialized so it is safe the unwrap
+    /// the lock without a check.
+    pub fn alloc(iface_id: String) {
+        if NET_DEV_METRICS_PVT
+            .read()
+            .unwrap()
+            .metrics
+            .get(&iface_id)
+            .is_none()
+        {
+            NET_DEV_METRICS_PVT
+                .write()
+                .unwrap()
+                .metrics
+                .insert(iface_id, NetDeviceMetrics::default());
+        }
+    }
+}
+
+#[macro_export]
+// Per device metrics are behind a ref lock and in a map which
+// makes it difficult to return the metrics pointer via a function
+// This macro is an attempt to make the code more readable and to
+// centralize the error handling.
+// The macro can be called in below ways where iface_id is String and
+// activate_fails is a metric from struct `NetDeviceMetrics`:
+//     net_metrics!(&iface_id, activate_fails.inc());
+//     net_metrics!(&iface_id, activate_fails.add(5));
+//     net_metrics!(&iface_id, activate_fails.count());
+macro_rules! net_metrics {
+    ($iface_id:expr,$metric:ident.$action:ident($($value:tt)?)) => {{
+        if $crate::devices::virtio::net::metrics::get_net_metrics().metrics.get($iface_id).is_some() {
+            $crate::devices::virtio::net::metrics::get_net_metrics().metrics.get($iface_id).unwrap().$metric.$action($($value)?)
+        }else{
+            $crate::devices::virtio::net::metrics::NetMetricsPerDevice::alloc($iface_id.to_string());
+            $crate::devices::virtio::net::metrics::get_net_metrics().metrics.get($iface_id).unwrap().$metric.$action($($value)?)
+        }
+    }}
+}
```

Comment:
Can't this (and `get_net_metrics` from above) be replaced by:
```rust
fn get_net_metrics(id: &str) -> &NetDeviceMetrics {
  NET_DEV_METRICS_PVT.read().unwrap().get(id).unwrap()
}
```
Because there should never be a situation when we created a device, but did not create a metrics for it.

Or maybe just add `get_metrics` method to the `Net` itself:
```rust
fn get_metrics(&self) -> &NetDeviceMetrics {
  NET_DEV_METRICS_PVT.read().unwrap().get(self.id).unwrap()
}
```
Even though it needs `NET_DEV_METRICS_PVT` to  be public, it is not a big problem.

Reply:
I am working on Patrick's suggestion of Arc which will remove the need for the macro net_metrics and get_net_metrics.

Reply:
both get_metrics and the macro are removed.